File: apollo-compiler/src/test/graphql/com/example/mutation_create_review_semantic_naming/CreateReviewForEpisodeMutation.java
package com.example.mutation_create_review_semantic_naming;

import com.apollographql.apollo.api.InputFieldMarshaller;
import com.apollographql.apollo.api.InputFieldWriter;
import com.apollographql.apollo.api.Mutation;
import com.apollographql.apollo.api.Operation;
import com.apollographql.apollo.api.OperationName;
import com.apollographql.apollo.api.ResponseField;
import com.apollographql.apollo.api.ResponseFieldMapper;
import com.apollographql.apollo.api.ResponseFieldMarshaller;
import com.apollographql.apollo.api.ResponseReader;
import com.apollographql.apollo.api.ResponseWriter;
import com.apollographql.apollo.api.internal.Optional;
import com.apollographql.apollo.api.internal.UnmodifiableMapBuilder;
import com.apollographql.apollo.api.internal.Utils;
import com.example.mutation_create_review_semantic_naming.type.Episode;
import com.example.mutation_create_review_semantic_naming.type.ReviewInput;
import java.io.IOException;
import java.lang.Object;
import java.lang.Override;
import java.lang.String;
import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.Map;
import javax.annotation.Generated;
import javax.annotation.Nonnull;
import javax.annotation.Nullable;

@Generated("Apollo GraphQL")
public final class CreateReviewForEpisodeMutation implements Mutation<CreateReviewForEpisodeMutation.Data, Optional<CreateReviewForEpisodeMutation.Data>, CreateReviewForEpisodeMutation.Variables> {
  public static final String OPERATION_DEFINITION = "mutation CreateReviewForEpisode($ep: Episode!, $review: ReviewInput!) {\n"
      + "  createReview(episode: $ep, review: $review) {\n"
      + "    __typename\n"
      + "    stars\n"
      + "    commentary\n"
      + "  }\n"
      + "}";

  public static final String QUERY_DOCUMENT = OPERATION_DEFINITION;

  private static final OperationName OPERATION_NAME = new OperationName() {
    @Override
    public String name() {
      return "CreateReviewForEpisode";
    }
  };

  private final CreateReviewForEpisodeMutation.Variables variables;

  public CreateReviewForEpisodeMutation(@Nonnull Episode ep, @Nonnull ReviewInput review) {
    Utils.checkNotNull(ep, "ep == null");
    Utils.checkNotNull(review, "review == null");
    variables = new CreateReviewForEpisodeMutation.Variables(ep, review);
  }

  @Override
  public String operationId() {
    return "eb015fa9dd6e305a9228393e61579154ae22719f6a18df6d00b45659ee2e7f7f";
  }

  @Override
  public String queryDocument() {
    return QUERY_DOCUMENT;
  }

  @Override
  public Optional<CreateReviewForEpisodeMutation.Data> wrapData(CreateReviewForEpisodeMutation.Data data) {
    return Optional.fromNullable(data);
  }

  @Override
  public CreateReviewForEpisodeMutation.Variables variables() {
    return variables;
  }

  @Override
  public ResponseFieldMapper<CreateReviewForEpisodeMutation.Data> responseFieldMapper() {
    return new Data.Mapper();
  }

  public static Builder builder() {
    return new Builder();
  }

  @Override
  public OperationName name() {
    return OPERATION_NAME;
  }

  public static final class Builder {
    private @Nonnull Episode ep;

    private @Nonnull ReviewInput review;

    Builder() {
    }

    public Builder ep(@Nonnull Episode ep) {
      this.ep = ep;
      return this;
    }

    public Builder review(@Nonnull ReviewInput review) {
      this.review = review;
      return this;
    }

    public CreateReviewForEpisodeMutation build() {
      Utils.checkNotNull(ep, "ep == null");
      Utils.checkNotNull(review, "review == null");
      return new CreateReviewForEpisodeMutation(ep, review);
    }
  }

  public static final class Variables extends Operation.Variables {
    private final @Nonnull Episode ep;

    private final @Nonnull ReviewInput review;

    private final transient Map<String, Object> valueMap = new LinkedHashMap<>();

    Variables(@Nonnull Episode ep, @Nonnull ReviewInput review) {
      this.ep = ep;
      this.review = review;
      this.valueMap.put("ep", ep);
      this.valueMap.put("review", review);
    }

    public @Nonnull Episode ep() {
      return ep;
    }

    public @Nonnull ReviewInput review() {
      return review;
    }

    @Override
    public Map<String, Object> valueMap() {
      return Collections.unmodifiableMap(valueMap);
    }

    @Override
    public InputFieldMarshaller marshaller() {
      return new InputFieldMarshaller() {
        @Override
        public void marshal(InputFieldWriter writer) throws IOException {
          writer.writeString("ep", ep.name());
          writer.writeObject("review", review.marshaller());
        }
      };
    }
  }

  public static class Data implements Operation.Data {
    static final ResponseField[] $responseFields = {
      ResponseField.forObject("createReview", "createReview", new UnmodifiableMapBuilder<String, Object>(2)
        .put("review", new UnmodifiableMapBuilder<String, Object>(2)
          .put("kind", "Variable")
          .put("variableName", "review")
        .build())
        .put("episode", new UnmodifiableMapBuilder<String, Object>(2)
          .put("kind", "Variable")
          .put("variableName", "ep")
        .build())
      .build(), true, Collections.<ResponseField.Condition>emptyList())
    };

    final Optional<CreateReview> createReview;

    private volatile String $toString;

    private volatile int $hashCode;

    private volatile boolean $hashCodeMemoized;

    public Data(@Nullable CreateReview createReview) {
      this.createReview = Optional.fromNullable(createReview);
    }

    public Optional<CreateReview> createReview() {
      return this.createReview;
    }

    public ResponseFieldMarshaller marshaller() {
      return new ResponseFieldMarshaller() {
        @Override
        public void marshal(ResponseWriter writer) {
          writer.writeObject($responseFields[0], createReview.isPresent() ? createReview.get().marshaller() : null);
        }
      };
    }

    @Override
    public String toString() {
      if ($toString == null) {
        $toString = "Data{"
          + "createReview=" + createReview
          + "}";
      }
      return $toString;
    }

    @Override
    public boolean equals(Object o) {
      if (o == this) {
        return true;
      }
      if (o instanceof Data) {
        Data that = (Data) o;
        return this.createReview.equals(that.createReview);
      }
      return false;
    }

    @Override
    public int hashCode() {
      if (!$hashCodeMemoized) {
        int h = 1;
        h *= 1000003;
        h ^= createReview.hashCode();
        $hashCode = h;
        $hashCodeMemoized = true;
      }
      return $hashCode;
    }

    public static final class Mapper implements ResponseFieldMapper<Data> {
      final CreateReview.Mapper createReviewFieldMapper = new CreateReview.Mapper();

      @Override
      public Data map(ResponseReader reader) {
        final CreateReview createReview = reader.readObject($responseFields[0], new ResponseReader.ObjectReader<CreateReview>() {
          @Override
          public CreateReview read(ResponseReader reader) {
            return createReviewFieldMapper.map(reader);
          }
        });
        return new Data(createReview);
      }
    }
  }

  public static class CreateReview {
    static final ResponseField[] $responseFields = {
      ResponseField.forString("__typename", "__typename", null, false, Collections.<ResponseField.Condition>emptyList()),
      ResponseField.forInt("stars", "stars", null, false, Collections.<ResponseField.Condition>emptyList()),
      ResponseField.forString("commentary", "commentary", null, true, Collections.<ResponseField.Condition>emptyList())
    };

    final @Nonnull String __typename;

    final int stars;

    final Optional<String> commentary;

    private volatile String $toString;

    private volatile int $hashCode;

    private volatile boolean $hashCodeMemoized;

    public CreateReview(@Nonnull String __typename, int stars, @Nullable String commentary) {
      this.__typename = Utils.checkNotNull(__typename, "__typename == null");
      this.stars = stars;
      this.commentary = Optional.fromNullable(commentary);
    }

    public @Nonnull String __typename() {
      return this.__typename;
    }

    /**
     * The number of stars this review gave, 1-5
     */
    public int stars() {
      return this.stars;
    }

    /**
     * Comment about the movie
     */
    public Optional<String> commentary() {
      return this.commentary;
    }

    public ResponseFieldMarshaller marshaller() {
      return new ResponseFieldMarshaller() {
        @Override
        public void marshal(ResponseWriter writer) {
          writer.writeString($responseFields[0], __typename);
          writer.writeInt($responseFields[1], stars);
          writer.writeString($responseFields[2], commentary.isPresent() ? commentary.get() : null);
        }
      };
    }

    @Override
    public String toString() {
      if ($toString == null) {
        $toString = "CreateReview{"
          + "__typename=" + __typename + ", "
          + "stars=" + stars + ", "
          + "commentary=" + commentary
          + "}";
      }
      return $toString;
    }

    @Override
    public boolean equals(Object o) {
      if (o == this) {
        return true;
      }
      if (o instanceof CreateReview) {
        CreateReview that = (CreateReview) o;
        return this.__typename.equals(that.__typename)
         && this.stars == that.stars
         && this.commentary.equals(that.commentary);
      }
      return false;
    }

    @Override
    public int hashCode() {
      if (!$hashCodeMemoized) {
        int h = 1;
        h *= 1000003;
        h ^= __typename.hashCode();
        h *= 1000003;
        h ^= stars;
        h *= 1000003;
        h ^= commentary.hashCode();
        $hashCode = h;
        $hashCodeMemoized = true;
      }
      return $hashCode;
    }

    public static final class Mapper implements ResponseFieldMapper<CreateReview> {
      @Override
      public CreateReview map(ResponseReader reader) {
        final String __typename = reader.readString($responseFields[0]);
        final int stars = reader.readInt($responseFields[1]);
        final String commentary = reader.readString($responseFields[2]);
        return new CreateReview(__typename, stars, commentary);
      }
    }
  }
}

File: apollo-compiler/src/test/graphql/com/example/mutation_create_review/CreateReviewForEpisodeMutation.kt
package com.example.mutation_create_review

import com.apollographql.apollo.api.InputFieldMarshaller
import com.apollographql.apollo.api.InputFieldWriter
import com.apollographql.apollo.api.Mutation
import com.apollographql.apollo.api.Operation
import com.apollographql.apollo.api.OperationName
import com.apollographql.apollo.api.ResponseField
import com.apollographql.apollo.api.ResponseFieldMapper
import com.apollographql.apollo.api.ResponseFieldMarshaller
import com.apollographql.apollo.api.ResponseReader
import com.example.mutation_create_review.type.CustomType
import com.example.mutation_create_review.type.Episode
import com.example.mutation_create_review.type.ReviewInput
import java.io.IOException
import java.util.Date
import javax.annotation.Generated
import kotlin.Any
import kotlin.Array
import kotlin.Int
import kotlin.String
import kotlin.Suppress
import kotlin.collections.List
import kotlin.collections.Map
import kotlin.jvm.Throws
import kotlin.jvm.Transient

@Generated("Apollo GraphQL")
data class CreateReviewForEpisodeMutation(val ep: Episode, val review: ReviewInput) : Mutation<CreateReviewForEpisodeMutation.Data, CreateReviewForEpisodeMutation.Data, Operation.Variables> {
    @Transient
    private val variables: Operation.Variables = object : Operation.Variables() {
        override fun valueMap(): Map<String, Any?> = mutableMapOf<String, Any?>().apply {
            this["ep"] = ep
            this["review"] = review
        }

        override fun marshaller(): InputFieldMarshaller = object : InputFieldMarshaller {
            @Throws(IOException::class)
            override fun marshal(writer: InputFieldWriter) {
                writer.writeString("ep", ep.value.rawValue)
                writer.writeObject("review", review.marshaller())
            }
        }
    }

    override fun operationId(): String = OPERATION_ID
    override fun queryDocument(): String = QUERY_DOCUMENT
    override fun wrapData(data: CreateReviewForEpisodeMutation.Data): CreateReviewForEpisodeMutation.Data = data
    override fun variables(): Operation.Variables = variables
    override fun name(): OperationName = OPERATION_NAME
    override fun responseFieldMapper(): ResponseFieldMapper<CreateReviewForEpisodeMutation.Data> = ResponseFieldMapper {
        CreateReviewForEpisodeMutation.Data(it)
    }

    /**
     * @param name The name of the character
     */
    data class ListOfListOfObject(
        val __typename: String,
        val name: String
    ) {
        fun marshaller(): ResponseFieldMarshaller = ResponseFieldMarshaller {
            it.writeString(RESPONSE_FIELDS[0], __typename)
            it.writeString(RESPONSE_FIELDS[1], name)
        }

        companion object {
            private val RESPONSE_FIELDS: Array<ResponseField> = arrayOf(
                    ResponseField.forString("__typename", "__typename", null, false, null),
                    ResponseField.forString("name", "name", null, false, null)
                    )

            operator fun invoke(reader: ResponseReader): ListOfListOfObject {
                val __typename = reader.readString(RESPONSE_FIELDS[0])
                val name = reader.readString(RESPONSE_FIELDS[1])
                return ListOfListOfObject(
                    __typename = __typename,
                    name = name
                )
            }
        }
    }

    /**
     * @param stars The number of stars this review gave, 1-5
     * @param commentary Comment about the movie
     * @param listOfListOfString for test purpose only
     * @param listOfListOfEnum for test purpose only
     * @param listOfListOfCustom for test purpose only
     * @param listOfListOfObject for test purpose only
     */
    data class CreateReview(
        val __typename: String,
        val stars: Int,
        val commentary: String?,
        val listOfListOfString: List<List<String>>?,
        val listOfListOfEnum: List<List<Episode>>?,
        val listOfListOfCustom: List<List<Date>>?,
        val listOfListOfObject: List<List<ListOfListOfObject>>?
    ) {
        fun marshaller(): ResponseFieldMarshaller = ResponseFieldMarshaller {
            it.writeString(RESPONSE_FIELDS[0], __typename)
            it.writeInt(RESPONSE_FIELDS[1], stars)
            it.writeString(RESPONSE_FIELDS[2], commentary)
            it.writeList(RESPONSE_FIELDS[3], listOfListOfString) { value, listItemWriter ->
                @Suppress("NAME_SHADOWING")
                value?.forEach { value ->
                    listItemWriter.writeList(value) { value, listItemWriter ->
                        value?.forEach { value ->
                            listItemWriter.writeString(value)
                        }
                    }
                }
            }
            it.writeList(RESPONSE_FIELDS[4], listOfListOfEnum) { value, listItemWriter ->
                @Suppress("NAME_SHADOWING")
                value?.forEach { value ->
                    listItemWriter.writeList(value) { value, listItemWriter ->
                        value?.forEach { value ->
                            listItemWriter.writeString(value)
                        }
                    }
                }
            }
            it.writeList(RESPONSE_FIELDS[5], listOfListOfCustom) { value, listItemWriter ->
                @Suppress("NAME_SHADOWING")
                value?.forEach { value ->
                    listItemWriter.writeList(value) { value, listItemWriter ->
                        value?.forEach { value ->
                            listItemWriter.writeCustom(CustomType.DATE, value)
                        }
                    }
                }
            }
            it.writeList(RESPONSE_FIELDS[6], listOfListOfObject) { value, listItemWriter ->
                @Suppress("NAME_SHADOWING")
                value?.forEach { value ->
                    listItemWriter.writeList(value) { value, listItemWriter ->
                        value?.forEach { value ->
                            listItemWriter.writeObject(value?.marshaller())
                        }
                    }
                }
            }
        }

        companion object {
            private val RESPONSE_FIELDS: Array<ResponseField> = arrayOf(
                    ResponseField.forString("__typename", "__typename", null, false, null),
                    ResponseField.forInt("stars", "stars", null, false, null),
                    ResponseField.forString("commentary", "commentary", null, true, null),
                    ResponseField.forList("listOfListOfString", "listOfListOfString", null, true, null),
                    ResponseField.forList("listOfListOfEnum", "listOfListOfEnum", null, true, null),
                    ResponseField.forList("listOfListOfCustom", "listOfListOfCustom", null, true, null),
                    ResponseField.forList("listOfListOfObject", "listOfListOfObject", null, true, null)
                    )

            operator fun invoke(reader: ResponseReader): CreateReview {
                val __typename = reader.readString(RESPONSE_FIELDS[0])
                val stars = reader.readInt(RESPONSE_FIELDS[1])
                val commentary = reader.readString(RESPONSE_FIELDS[2])
                val listOfListOfString = reader.readList<List<String>>(RESPONSE_FIELDS[3]) {
                    it.readList<String> {
                        it.readString()
                    }
                }
                val listOfListOfEnum = reader.readList<List<Episode>>(RESPONSE_FIELDS[4]) {
                    it.readList<Episode> {
                        Episode.safeValueOf(it.readString())
                    }
                }
                val listOfListOfCustom = reader.readList<List<Date>>(RESPONSE_FIELDS[5]) {
                    it.readList<Date> {
                        it.readCustomType<Date>(CustomType.DATE)
                    }
                }
                val listOfListOfObject = reader.readList<List<ListOfListOfObject>>(RESPONSE_FIELDS[6]) {
                    it.readList<ListOfListOfObject> {
                        it.readObject<ListOfListOfObject> {
                            ListOfListOfObject(it)
                        }

                    }
                }
                return CreateReview(
                    __typename = __typename,
                    stars = stars,
                    commentary = commentary,
                    listOfListOfString = listOfListOfString,
                    listOfListOfEnum = listOfListOfEnum,
                    listOfListOfCustom = listOfListOfCustom,
                    listOfListOfObject = listOfListOfObject
                )
            }
        }
    }

    data class Data(val createReview: CreateReview?) : Operation.Data {
        override fun marshaller(): ResponseFieldMarshaller = ResponseFieldMarshaller {
            it.writeObject(RESPONSE_FIELDS[0], createReview?.marshaller())
        }

        companion object {
            private val RESPONSE_FIELDS: Array<ResponseField> = arrayOf(
                    ResponseField.forObject("createReview", "createReview", mapOf<String, Any>(
                        "episode" to mapOf<String, Any>(
                            "kind" to "Variable",
                            "variableName" to "ep"),
                        "review" to mapOf<String, Any>(
                            "kind" to "Variable",
                            "variableName" to "review")), true, null)
                    )

            operator fun invoke(reader: ResponseReader): Data {
                val createReview = reader.readObject<CreateReview>(RESPONSE_FIELDS[0]) {
                    CreateReview(it)
                }

                return Data(
                    createReview = createReview
                )
            }
        }
    }

    companion object {
        val OPERATION_DEFINITION: String = """
                |mutation CreateReviewForEpisode(${'$'}ep: Episode!, ${'$'}review: ReviewInput!) {
                |  createReview(episode: ${'$'}ep, review: ${'$'}review) {
                |    __typename
                |    stars
                |    commentary
                |    listOfListOfString
                |    listOfListOfEnum
                |    listOfListOfCustom
                |    listOfListOfObject {
                |      __typename
                |      name
                |    }
                |  }
                |}
                """.trimMargin()

        const val OPERATION_ID: String =
                "63f677799b4c5cfe580f133b71712c66179ea84077be680e999b0f1d1e66e0a1"

        val QUERY_DOCUMENT: String = """
                |mutation CreateReviewForEpisode(${'$'}ep: Episode!, ${'$'}review: ReviewInput!) {
                |  createReview(episode: ${'$'}ep, review: ${'$'}review) {
                |    __typename
                |    stars
                |    commentary
                |    listOfListOfString
                |    listOfListOfEnum
                |    listOfListOfCustom
                |    listOfListOfObject {
                |      __typename
                |      name
                |    }
                |  }
                |}
                """.trimMargin()

        val OPERATION_NAME: OperationName = OperationName { "CreateReviewForEpisodeMutation" }
    }
}
